tracks/javascript/exercises/leap/mentoring.md

```diff
@@ -5,71 +5,186 @@ The point of this exercise is to teach the student to use:
+- If a student submits one-liner with a `return`, talk about implicit return of arrow functions with
+  [concise body rather than block body](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Function_body).
+
+#### Maintainability and performance:
+Something like this ís readable and maintainable, but if a student submits this, point out resources (functions are created each invocation) and returning expression evaluations as in this example, each invocation `isDivisibleFn` is recreated, which can be fixed by inversing the curry; and all values are calculated before they are used. In fact, when `by400` is true, the rest doesn't matter, so it's a waste of CPU cycles. Applying the suggestions to this works the same when it comes to logical operators, _order of operations_ and _order of evaluation_ as per the second reasonable solution.
 
-  ```javascript
-  function isDivisibleBy(number) {
-    return mod => number % mod === 0;
+```javascript
+function isDivisibleBy(number) {
+  return mod => number % mod === 0;
+}
+
+export const isLeap = year => {
+  const isDivisibleByFn = isDivisibleBy(year);
+  const by4 = isDivisibleByFn(4);
+  const by100 = isDivisibleByFn(100);
+  const by400 = isDivisibleByFn(400);
+  let isLeapResult = Boolean(by4 && !by100);
+  if (by400) {
+    isLeapResult = true;
   }
+  return isLeapResult;
+};
+```
+
+#### Short circuit explanation
+Some students ask something along the lines of:
+
+> Are some expressions not evaluated if they won't effect the result?
 
-  export const isLeap = year => {
-    const isDivisibleByFn = isDivisibleBy(year);
-    const by4 = isDivisibleByFn(4);
-    const by100 = isDivisibleByFn(100);
-    const by400 = isDivisibleByFn(400);
-    let isLeapResult = Boolean(by4 && !by100);
-    if (by400) {
-      isLeapResult = true;
+
+Here is a template you can respond with:
+
+    That is a good question. In JavaScript and Typescript, some operators have [short-circuit](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_Operators#Short-circuit_evaluation) evaluation.
+
+    - `a && b`
+      - `a` is _always_ evaluated
+      - `b` is _only_ evaluated if `a` was `truthy`, otherwise the entire expression short-circuits to whatever `a` yielded.
+    - `a || b`
+      - `a` is _always_ evaluated
+      - `b` is _only_ evaluated if `b` was `falsy`, otherwise the entire expression short-circuits to whatever `a` yielded.
```

Comment:
`yielded` might sound slightly confusing, maybe use `evaluated`? Although both are same, ignore if you feel this is OK.